src/main/java/com/brightpath/learnify/domain/auth/PermissionAccessService.java

```diff
@@ -109,14 +113,63 @@ public void removeUserWithAccessToResource(UUID resourceId, ResourceType resourc
         PermissionsAccessEntity permissionsAccessEntity = permissionAccessRepository.findById(permissionAccessId(resourceId, resourceType))
                 .orElseThrow(() -> new ResourceNotFoundException(resourceType));
 
-        if(permissionsAccessEntity.getOwnerId().equals(userId)) {
+        if (permissionsAccessEntity.getOwnerId().equals(userId)) {
             throw new IllegalArgumentException("Cannot remove owner access");
         }
 
         permissionsAccessEntity.removePermissionForUser(userId);
         permissionAccessRepository.save(permissionsAccessEntity);
     }
 
+    public void checkUserPermissionToEditResource(UUID resourceId, ResourceType resourceType) {
+        String userId = userIdentityService.getCurrentUserId();
+        boolean hasAccessToEditNote = hasUserAccessToResource(userId, resourceId, resourceType, READ_WRITE);
+        if (!hasAccessToEditNote) {
+            throw new UserNotAuthorizedToEditException();
+        }
+    }
+
+    public void checkUserPermissionToViewResource(UUID resourceId, ResourceType resourceType) {
+        String userId = userIdentityService.getCurrentUserId();
+        boolean hasAccessToEditNote = hasUserAccessToResource(userId, resourceId, resourceType, READ_ONLY);
+        if (!hasAccessToEditNote) {
+            throw new UserNotAuthorizedToGetException();
+        }
+    }
+
+    public void addPermissionToResourceForUser(UUID resourceId, ResourceType resourceType, String userId, ResourceAccessEnum requestedAccess) {
+        if (userAnyHasPermissionToResource(resourceId, resourceType, userId, Optional.empty())) {
+            throw new UserAccessIsAlreadyGrantedException();
+        }
+        //todo save access to db
+    }
+
+    public void editPermissionToResourceForUser(UUID resourceId, ResourceType resourceType, String userId, ResourceAccessEnum requestedAccess) {
+        if (userAnyHasPermissionToResource(resourceId, resourceType, userId, Optional.of(requestedAccess))) {
+            throw new UserAccessIsAlreadyGrantedException();
+        }
+        //todo save access to db
+    }
+
+    public void deletePermissionToResourceForUser(UUID resourceId, ResourceType resourceType, String userId, ResourceAccessEnum requestedAccess) {
+        if (!userAnyHasPermissionToResource(resourceId, resourceType, userId, Optional.of(requestedAccess))) {
+            throw new UserDoesNotHavePermissionToRemoveException();
+        }
+        //todo remove access from db
+    }
+
+    private boolean userAnyHasPermissionToResource(UUID resourceId, ResourceType resourceType, String userId, Optional<ResourceAccessEnum> access) {
```

Comment:
I really don't understand what this method does :eyes:

Reply:
it check if user has permission (the one in opitonal) or if it's empty it checks if he has any

Reply:
i added two more to make it more clear, imo now its better